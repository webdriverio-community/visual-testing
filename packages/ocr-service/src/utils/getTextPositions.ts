import getData from './getData.js'
import { getDprPositions } from './index.js'
import type { Line, GetTextPositions, GetTextPositionsOptions } from '../types.js'

export default async function getTextPositions(browser: WebdriverIO.Browser, options: GetTextPositionsOptions): Promise<GetTextPositions[]> {
    const { dpr, filePath, words } = await getData(browser, options)

    return (
        words
            .map(({ text, bbox }:Line) => ({
                dprPosition: getDprPositions(JSON.parse(JSON.stringify(bbox)), dpr),
                filePath,
                originalPosition: bbox,
                text: text.replace(/(^\s+|\s+$)/g, ''),
            }))
            .filter((word) => word.text)
    )
}
