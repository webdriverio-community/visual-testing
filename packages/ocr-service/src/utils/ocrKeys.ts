import { Key } from 'webdriverio'

export default async function ocrKeys(value: string, submitValue: boolean): Promise<void> {
    const actionChain = browser.action('key');

    [...value].forEach((char: string) => {
        actionChain.down(char)
    });
    [...value].forEach((char: string) => {
        actionChain.up(char)
    })

    if (submitValue) {
        /**
         * XCTest API only allows to send keypresses (e.g. keydown+keyup).
         */
        if (!browser.isIOS){
            actionChain.pause(50)
        }
        actionChain.down(Key.Enter)
        actionChain.up(Key.Enter)
    }
    await actionChain.perform()
}
