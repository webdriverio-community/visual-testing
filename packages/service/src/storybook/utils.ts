import type { Logger } from '@wdio/logger'
import type { Capabilities, Options } from '@wdio/types'
import fetch from 'node-fetch'
import { mkdirSync, writeFileSync } from 'node:fs'
import { tmpdir } from 'node:os'
import { join, resolve } from 'node:path'
import type { ClassOptions } from 'webdriver-image-comparison'
import type {
    CategoryComponent,
    CreateItContent,
    CreateTestContent,
    CreateTestFileOptions,
    IndexRes,
    ScanStorybookReturnData,
    Stories,
    StoriesRes,
    StorybookData,
} from './Types.js'

/**
 * Check if we run for Storybook
 */
export function isStorybookMode(): boolean {
    return process.argv.includes('--storybook')
}

/**
 * Check if the framework is cucumber
 */
export function isCucumberFramework(framework: string): boolean {
    return framework.toLowerCase() === 'cucumber'
}

/**
 * Check if the framework is Jasmine
 */
export function isJasmineFramework(framework: string): boolean {
    return framework.toLowerCase() === 'jasmine'
}

/**
 * Check if the framework is Mocha
 */
export function isMochaFramework(framework: string): boolean {
    return framework.toLowerCase() === 'mocha'
}

/**
 * Check if there is an instance of Storybook running
 */
export async function checkStorybookIsRunning(url: string) {
    try {
        const res = await fetch(url, { method: 'GET', headers: {} })
        if (res.status !== 200) {
            throw new Error(`Unxpected status: ${res.status}`)
        }
    } catch (e) {
        console.error(`It seems that the Storybook instance is not running at: ${url}. Are you sure it's running?`)
        process.exit(1)
    }
}

/**
 * Sanitize the URL to ensure it's in a proper format
 */
export function sanitizeURL(url: string): string {
    if (!url.startsWith('http://') && !url.startsWith('https://')) {
        url = 'http://' + url
    }

    url = url.replace(/(iframe\.html|index\.html)\s*$/, '')

    if (!url.endsWith('/')) {
        url += '/'
    }

    return url
}

/**
 * Extract the category and component from the story ID
 */
export function extractCategoryAndComponent(id: string): CategoryComponent {
    // The ID is in the format of `category-component--storyName`
    const [categoryComponent] = id.split('--')
    const [category, component] = categoryComponent.split('-')

    return { category, component }
}

/**
 * Get the stories JSON from the Storybook instance
 */
export async function getStoriesJson(url: string): Promise<Stories> {
    const indexJsonUrl = new URL('index.json', url).toString()
    const storiesJsonUrl = new URL('stories.json', url).toString()
    const fetchOptions = { headers: {} }

    try {
        const [indexRes, storiesRes] = await Promise.all([
            fetch(indexJsonUrl, fetchOptions),
            fetch(storiesJsonUrl, fetchOptions),
        ])

        for (const response of [storiesRes, indexRes]) {
            if (response.ok) {
                const data = await response.json() as StoriesRes | IndexRes
                return (data as StoriesRes).stories || (data as IndexRes).entries
            }
        }
    } catch (err) {
        console.error(err)
    }

    throw new Error(`Failed to fetch index data from the project. Ensure URLs are available with valid data: ${storiesJsonUrl}, ${indexJsonUrl}.`)
}

/**
 * Get arg value from the process.argv
 */
export function getArgvValue(argName: string, parseFunc: (value: string) => any): any {
    const argWithEqual = argName + '='
    const argv = process.argv

    for (let i = 0; i < argv.length; i++) {
        if (argv[i] === argName && i + 1 < argv.length) {
            return parseFunc(argv[i + 1])
        } else if (argv[i].startsWith(argWithEqual)) {
            const value = argv[i].slice(argWithEqual.length)
            return parseFunc(value)
        }
    }

    return undefined
}

/**
 * Creates a it function for the test file
 * @TODO: improve this
 */
export function itFunction({ clip, clipSelector, folders: { baselineFolder }, framework, skipStories, storyData, storybookUrl }: CreateItContent) {
    const { id } = storyData
    const screenshotType = clip ? 'n element' : ' viewport'
    const DEFAULT_IT_TEXT = 'it'
    let itText = DEFAULT_IT_TEXT

    if (isJasmineFramework(framework)) {
        itText = 'xit'
    } else if (isMochaFramework(framework)) {
        itText = 'it.skip'
    }

    if (Array.isArray(skipStories)) {
        itText = skipStories.includes(id) ? itText : DEFAULT_IT_TEXT
    } else if (skipStories instanceof RegExp) {
        itText = skipStories.test(id) ? itText : DEFAULT_IT_TEXT
    }

    // Setup the folder structure
    const { category, component } = extractCategoryAndComponent(id)
    const methodOptions = {
        baselineFolder: join(baselineFolder, `./${category}/${component}/`),
    }

    const it = `
    ${itText}(\`should take a${screenshotType} screenshot of ${id}\`, async () => {
        await browser.url(\`${storybookUrl}iframe.html?id=${id}\`);
        await $('${clipSelector}').waitForDisplayed();
        await waitForAllImagesLoaded();
        ${clip
        ? `await expect($('${clipSelector}')).toMatchElementSnapshot('${id}-element', ${JSON.stringify(methodOptions)})`
        : `await expect(browser).toMatchScreenSnapshot('${id}', ${JSON.stringify(methodOptions)})`}
    });
    `
    return it
}

/**
 * Write the test file
 */
export function writeTestFile(directoryPath: string, fileID: string, log: Logger, testContent: string) {
    const filePath = join(directoryPath, `${fileID}.test.js`)
    try {
        writeFileSync(filePath, testContent)
        log.info(`Test file created at: ${filePath}`)
    } catch (err) {
        log.error(`It seems that the writing the file to '${filePath}' didn't succeed due to the following error: ${err}`)
        process.exit(1)
    }
}

/**
 * Create the test content
 */
export function createTestContent(
    { clip, clipSelector, folders, framework, skipStories, stories, storybookUrl }: CreateTestContent,
    // For testing purposes only
    itFunc = itFunction
): string {
    const itFunctionOptions = { clip, clipSelector, folders, framework, skipStories, storybookUrl }

    return stories.reduce((acc, storyData) => acc + itFunc({ ...itFunctionOptions, storyData }), '')
}

const waitForAllImagesLoaded = `
async function waitForAllImagesLoaded() {
    await browser.executeAsync(async (done) => {
        const timeout = 11000; // 11 seconds
        let timedOut = false;

        const timeoutPromise = new Promise((resolve, reject) => {
            setTimeout(() => {
                timedOut = true;
                reject('Timeout: Not all images loaded within 11 seconds');
            }, timeout);
        });

        const isImageLoaded = (img) => img.complete && img.naturalWidth > 0;

        // Check for <img> elements
        const imgElements = Array.from(document.querySelectorAll('img'));
        const imgPromises = imgElements.map(img => isImageLoaded(img) ? Promise.resolve() : new Promise(resolve => {
            img.onload = () => !timedOut && resolve();
            img.onerror = () => !timedOut && resolve();
        }));

        // Check for CSS background images
        const allElements = Array.from(document.querySelectorAll('*'));
        const bgImagePromises = allElements.map(el => {
            const bgImage = window.getComputedStyle(el).backgroundImage;
            if (bgImage && bgImage !== 'none' && bgImage.startsWith('url')) {
                const imageUrl = bgImage.slice(5, -2); // Extract URL from the 'url("")'
                const image = new Image();
                image.src = imageUrl;
                return isImageLoaded(image) ? Promise.resolve() : new Promise(resolve => {
                    image.onload = () => !timedOut && resolve();
                    image.onerror = () => !timedOut && resolve();
                });
            }
            return Promise.resolve();
        });

        try {
            await Promise.race([Promise.all([...imgPromises, ...bgImagePromises]), timeoutPromise]);
            done();
        } catch (error) {
            done(error);
        }
    });
}
`

/**
 * Create the file data
 */
export function createFileData(describeTitle: string, testContent: string): string {
    return `
describe(\`${describeTitle}\`, () => {
    ${testContent}
});
${waitForAllImagesLoaded}
`
}

/**
 * Create the test files
 */
export function createTestFiles(
    { clip, clipSelector, directoryPath, folders, framework, log, numShards, skipStories, storiesJson, storybookUrl }: CreateTestFileOptions,
    // For testing purposes only
    createTestCont = createTestContent,
    createFileD = createFileData,
    writeTestF = writeTestFile
) {
    const storiesArray = Object.values(storiesJson)
        // By default only keep the stories, not the docs
        .filter((storyData: StorybookData) => storyData?.type === 'story' || !storyData.parameters?.docsOnly)
    const fileNamePrefix = 'visual-storybook'
    const createTestContentData = { clip, clipSelector, folders, framework, skipStories, stories: storiesArray, storybookUrl }

    if (numShards === 1) {
        const testContent = createTestCont(createTestContentData)
        const fileData = createFileD('All stories', testContent)
        writeTestF(directoryPath, `${fileNamePrefix}-1-1`, log, fileData)
    } else {
        const totalStories = storiesArray.length
        const storiesPerShard = Math.ceil(totalStories / numShards)

        for (let shard = 0; shard < numShards; shard++) {
            const startIndex = shard * storiesPerShard
            const endIndex = Math.min(startIndex + storiesPerShard, totalStories)
            const shardStories = storiesArray.slice(startIndex, endIndex)
            const testContent = createTestCont({ ...createTestContentData, stories: shardStories })
            const fileId = `${fileNamePrefix}-${shard + 1}-${numShards}`
            const describeTitle = `Shard ${shard + 1} of ${numShards}`
            const fileData = createFileD(describeTitle, testContent)

            writeTestF(directoryPath, fileId, log, fileData)
        }
    }
}

/**
 * Create the storybook capabilities based on the specified browsers
 */
export function createStorybookCapabilities(capabilities: Capabilities.RemoteCapabilities, log: Logger) {
    const isHeadless = getArgvValue('--headless', value => value !== 'false') ?? true
    const browsers = getArgvValue('--browsers', (value: string) => value.split(',')) ?? ['chrome']

    if (Array.isArray(capabilities)) {
        const chromeCapability: WebdriverIO.Capabilities = {
            browserName: 'chrome',
            'goog:chromeOptions': {
                args: [
                    'disable-infobars',
                    ...(isHeadless ? ['--headless'] : []),
                ],
            },
            'wdio-ics:options': {
                logName: 'local-chrome',
            },
        }
        const firefoxCapability: WebdriverIO.Capabilities = {
            browserName: 'firefox',
            'moz:firefoxOptions': {
                args: [...(isHeadless ? ['-headless'] : []),]
            },
            'wdio-ics:options': {
                logName: 'local-firefox',
            },
        }
        const safariCapability: WebdriverIO.Capabilities = {
            browserName: 'safari',
            'wdio-ics:options': {
                logName: 'local-safari',
            },
        }
        const edgeCapability: WebdriverIO.Capabilities = {
            browserName: 'MicrosoftEdge',
            'ms:edgeOptions': {
                args: [...(isHeadless ? ['--headless'] : [])]
            },
            'wdio-ics:options': {
                logName: 'local-edge',
            },
        }
        interface CapabilityMap {
            chrome: typeof chromeCapability;
            firefox: typeof firefoxCapability;
            safari: typeof safariCapability;
            edge: typeof edgeCapability;
        }
        const capabilityMap: CapabilityMap = {
            chrome: chromeCapability,
            firefox: firefoxCapability,
            safari: safariCapability,
            edge: edgeCapability,
        }
        const newCapabilities = browsers
            .filter((browser:string): browser is keyof CapabilityMap => browser in capabilityMap)
            .map((browser:string) => capabilityMap[browser as keyof CapabilityMap])
        capabilities.length = 0
        // Add the new capability to the capabilities array
        capabilities.push(...newCapabilities)
    } else {
        log.error('The capabilities are not an array')
    }
}

/**
 * Scan the storybook instance
 */
export async function scanStorybook(
    config: Options.Testrunner,
    log: Logger,
    options: ClassOptions,
    // For testing purposes only
    getArgvVal = getArgvValue,
    checkStorybookIsRun = checkStorybookIsRunning,
    sanitizeURLFunc = sanitizeURL,
    getStoriesJsonFunc = getStoriesJson,
): Promise<ScanStorybookReturnData> {
    // Prepare storybook scanning
    const cliUrl = getArgvVal('--url', value => value)
    const rawStorybookUrl = cliUrl ?? process.env.STORYBOOK_URL ?? options?.storybook?.url ?? 'http://127.0.0.1:6006'
    await checkStorybookIsRun(rawStorybookUrl)
    const storybookUrl = sanitizeURLFunc(rawStorybookUrl)

    // Create a temporary folder for test files and add that to the specs
    const tempDir = resolve(tmpdir(), `wdio-storybook-tests-${Date.now()}`)
    mkdirSync(tempDir)
    log.info(`Using temporary folder for storybook specs: ${tempDir}`)
    config.specs = [join(tempDir, '*.js')]

    // Get the stories
    const storiesJson = await getStoriesJsonFunc(storybookUrl)

    return {
        storiesJson,
        storybookUrl,
        tempDir,
    }
}

/**
 * Parse the stories to skip
 */
export function parseSkipStories(skipStories: string | string[], log: Logger): RegExp | string[] {
    if (Array.isArray(skipStories)) {
        return skipStories
    }

    const regexPattern = /^\/.*\/[gimyus]*$/
    if (regexPattern.test(skipStories)) {
        try {
            const match = skipStories.match(/^\/(.+)\/([gimyus]*)$/)
            if (match) {
                const [, pattern, flags] = match
                return new RegExp(pattern, flags)
            }
        } catch (error) {
            log.error('Invalid regular expression:', error, '. Not using a regular expression to skip stories.')
        }
    }

    return skipStories.split(',').map(skipped => skipped.trim())
}
