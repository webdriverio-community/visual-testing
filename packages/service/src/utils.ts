/// <reference types="webdriverio" />

import type { Capabilities } from '@wdio/types'
import type {
    Folders,
    InstanceData,
    CheckScreenMethodOptions,
    SaveScreenMethodOptions,
    CheckFullPageMethodOptions,
    SaveFullPageMethodOptions,
    CheckElementMethodOptions,
    SaveElementMethodOptions,
} from 'webdriver-image-comparison'

interface WdioIcsOptions {
    logName?: string;
    name?: string;
}

/**
 * Get the folders data
 *
 * If folder options are passed in use those values
 * Otherwise, use the values set during instantiation
 */
type getFolderMethodOptions =
    | CheckElementMethodOptions
    | CheckFullPageMethodOptions
    | CheckScreenMethodOptions
    | SaveElementMethodOptions
    | SaveFullPageMethodOptions
    | SaveScreenMethodOptions;
export function getFolders(
    methodOptions: getFolderMethodOptions,
    folders: Folders
): Folders {
    return {
        actualFolder: methodOptions.actualFolder ?? folders.actualFolder,
        baselineFolder: methodOptions.baselineFolder ?? folders.baselineFolder,
        diffFolder: methodOptions.diffFolder ?? folders.diffFolder,
    }
}

/**
 * Get the instance data
 */
export function getInstanceData(
    capabilities: WebdriverIO.Capabilities,
    currentBrowser: WebdriverIO.Browser
): InstanceData {
    const currentCapabilities = (currentBrowser.requestedCapabilities as Capabilities.W3CCapabilities).alwaysMatch
        ? (currentBrowser.requestedCapabilities as Capabilities.W3CCapabilities).alwaysMatch
        : (currentBrowser.requestedCapabilities as WebdriverIO.Capabilities)
    const browserName = (
        capabilities.browserName ||
        currentCapabilities.browserName ||
        'browserName-not-known'
    ).toLowerCase()
    const browserVersion = (
        capabilities.browserVersion ||
        currentCapabilities.browserVersion ||
        'not-known'
    ).toLowerCase()
    const logName =
        'wdio-ics:options' in capabilities
            ? (capabilities['wdio-ics:options'] as WdioIcsOptions)?.logName ??
              ''
            : ''
    const name =
        'wdio-ics:options' in capabilities
            ? (capabilities['wdio-ics:options'] as WdioIcsOptions)?.name ?? ''
            : ''

    // For mobile
    const platformName = (
        capabilities.platformName ||
        currentCapabilities.platformName ||
        'not-known'
    ).toLowerCase()
    const platformVersion = (
        capabilities['appium:platformVersion'] ||
        (currentCapabilities as Capabilities.AppiumCapabilities)[
            'appium:platformVersion'
        ] ||
        'not-known'
    ).toLowerCase()
    const deviceName = (capabilities['appium:deviceName'] || '').toLowerCase()
    const nativeWebScreenshot = !!(
        (capabilities as Capabilities.AppiumAndroidCapabilities)[
            'appium:nativeWebScreenshot'
        ] ||
        (currentCapabilities as Capabilities.AppiumAndroidCapabilities)[
            'appium:nativeWebScreenshot'
        ]
    )

    return {
        browserName,
        browserVersion,
        deviceName,
        logName,
        name,
        nativeWebScreenshot,
        platformName,
        platformVersion,
    }
}

/**
 * traverse up the scope chain until browser element was reached
 */
export function getBrowserObject (elem: WebdriverIO.Element | WebdriverIO.Browser): WebdriverIO.Browser {
    const elemObject = elem as WebdriverIO.Element
    return (elemObject as WebdriverIO.Element).parent ? getBrowserObject(elemObject.parent) : elem as WebdriverIO.Browser
}
