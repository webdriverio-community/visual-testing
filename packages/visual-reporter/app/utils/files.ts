export const getRelativePath = (fullPath: string): string => {
    const basePath = import.meta.env.BASE_URL || '/'
    const publicPathIndex = fullPath.indexOf('/public/')
    if (publicPathIndex !== -1) {
        return fullPath.slice(publicPathIndex + '/public/'.length)
    }

    return `${basePath}${fullPath}`
}
