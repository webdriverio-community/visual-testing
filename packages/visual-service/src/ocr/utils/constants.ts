export const SUPPORTED_LANGUAGES = {
    AFRIKAANS: 'afr',
    AMHARIC: 'amh',
    ARABIC: 'ara',
    ASSANESE: 'asm',
    AZERBAIJANI: 'aze',
    BELARUSIAN: 'bel',
    BENGALI: 'bel',
    TIBETAN: 'bod',
    BOSNIAN: 'bos',
    BULGARIAN: 'bul',
    CATALAN: 'cat',
    CZECH: 'cat',
    CHINESE: 'chi_sim',
    CORSICAN: 'cos',
    DANISH: 'dan',
    GERMAN: 'deu',
    GREEK: 'ell',
    ENGLISH: 'eng',
    MATH: 'equ',
    ESTONIAN: 'est',
    FINNISH: 'fin',
    FRENCH: 'fra',
    GAELIC: 'gla',
    IRISH: 'gle',
    HINDI: 'hin',
    CROATIAN: 'hrv',
    HUNGARIAN: 'hun',
    ARMENIAN: 'hye',
    ITALIAN: 'ita',
    JAPANESE: 'jpn',
    KANNADA: 'kan',
    KOREAN: 'kor',
    LATIN: 'lat',
    LATVIAN: 'lav',
    LITHUANIAN: 'lit',
    LUXEMBOURGISH: 'ltz',
    MACEDONIAN: 'mkd',
    DUTCH: 'nld',
    NORWEGIAN: 'nor',
    POLISH: 'pol',
    PORTUGUESE: 'por',
    ROMANIAN: 'ron',
    RUSSIAN: 'rus',
    SLOVAK: 'slk',
    SLOVENIAN: 'slv',
    SPANISH: 'spa',
    ALBANIAN: 'sqi',
    SERBIAN: 'srp',
    SWAHILI: 'swa',
    SWEDISH: 'swe',
    THAI: 'tha',
    TURKISH: 'tur',
    UKRAINIAN: 'ukr',
    VIETNAMESE: 'vie'
}
export const CONTRAST = 0.25
export const DEFAULT_FUZZY_OPTIONS = {
    distance: 100,
    isCaseSensitive: false,
    findAllMatches: false,
    location: 0,
    minMatchCharLength: 2,
    threshold: 0.6,
}
