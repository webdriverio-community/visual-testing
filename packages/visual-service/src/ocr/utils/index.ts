import { dirname, join } from 'node:path'
import { mkdirSync } from 'node:fs'
import type { Folders } from 'webdriver-image-comparison'
import type { ClickPoint, DetermineClickPointOptions, OcrOptions, Rectangles, RectReturn } from '../types.js'

export function getDprPositions(values: Rectangles, dpr: number): Rectangles {
    Object.keys({ ...values }).map((value: string) => {
    // @ts-ignore
        values[value] /= dpr
    })

    return values
}

/**
 * Determine the click point
 */
export function determineClickPoint(options: DetermineClickPointOptions): ClickPoint {
    const { rectangles: { left, right, top, bottom } } = options
    const x = Math.round( left + (right - left) / 2 )
    const y = Math.round( top + (bottom - top) / 2 )

    return { x, y }
}

/**
 * Adjust the Element Bbox
 */
export function adjustElementBbox(bbox: Rectangles, elementRect: RectReturn): Rectangles {
    return {
        left: Math.round(bbox.left + elementRect.x),
        top: Math.round(bbox.top + elementRect.y),
        right: Math.round(bbox.right + elementRect.x),
        bottom: Math.round(bbox.bottom + elementRect.y),
    }
}

export function createOcrDir(options: OcrOptions, folders: Folders): string {
    const ocrDir = options.ocr?.imagesPath || join(dirname(folders.actualFolder), 'ocr')
    mkdirSync(ocrDir, { recursive: true })

    return ocrDir
}

export function isRectanglesObject(obj: WebdriverIO.Element | ChainablePromiseElement | RectReturn): boolean {
    const properties = ['x', 'y', 'width', 'height']

    return obj !== null &&
           typeof obj === 'object' &&
           properties.every(prop => typeof (obj as any)[prop] === 'number' && !isNaN((obj as any)[prop]) && (obj as any)[prop] >= 0)
}
