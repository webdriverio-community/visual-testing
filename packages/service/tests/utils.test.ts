import { describe, it, expect } from 'vitest'
import { getFolders, getInstanceData } from '../src/utils.js'

describe('utils', () => {
    describe('getFolders', () => {
        it('should be able to return the correct folders when no method options are provided', () => {
            const methodOptions = {}
            const folders = {
                baselineFolder: 'folderBase',
                diffFolder: 'folderDiff',
                actualFolder: 'folderActual',
            }
            expect(getFolders(methodOptions, folders)).toMatchSnapshot()
        })

        it('should be able to return the correct folders when method options are provided', () => {
            const methodOptions = {
                baselineFolder: 'methodBase',
                diffFolder: 'methodDiff',
                actualFolder: 'methodActual',
            }
            const folders = {
                baselineFolder: 'folderBase',
                diffFolder: 'folderDiff',
                actualFolder: 'folderActual',
            }
            expect(getFolders(methodOptions, folders)).toMatchSnapshot()
        })
    })

    describe('getInstanceData', () => {
        const capabilities = {
            browserName: 'chrome',
            version: '75.123',
            platformName: 'osx',
            platformVersion: '12',
        } as WebdriverIO.Capabilities
        const browser = {
            capabilities,
            requestedCapabilities: capabilities,
        } as any as WebdriverIO.Browser

        it('should return instance data when the minimum of capabilities is provided', async() => {
            const capabilities = {} as WebdriverIO.Capabilities
            expect(await getInstanceData(capabilities, browser)).toMatchSnapshot()
        })

        it('should return instance data when wdio-ics option log name is provided', async() => {
            const capabilities = {
                browserName: 'chrome',
                'wdio-ics:options': {
                    logName: 'wdio-ics-logName',
                },
            } as WebdriverIO.Capabilities
            expect(await getInstanceData(capabilities, browser)).toMatchSnapshot()
        })

        it('should return instance data when log name is provided', async () => {
            const capabilities = {
                browserName: 'chrome',
                logName: 'logName',
                'wdio-ics:options': {
                    foo: 'bar',
                },
            } as WebdriverIO.Capabilities
            expect(await getInstanceData(capabilities, browser)).toMatchSnapshot()
        })

        it('should return instance data when the sauce log name is provided', async() => {
            const capabilities = {
                browserName: 'chrome',
                'sauce:options': {
                    logName: 'sauceLogName',
                },
            } as WebdriverIO.Capabilities
            expect(await getInstanceData(capabilities, browser)).toMatchSnapshot()
        })

        it('should return instance data when the appium log name is provided', async() => {
            const capabilities = {
                browserName: 'chrome',
                'appium:options': {
                    logName: 'appiumLogName',
                },
            } as WebdriverIO.Capabilities
            expect(await getInstanceData(capabilities, browser)).toMatchSnapshot()
        })

        it('should return instance data when all capabilities are provided', async() => {
            const capabilities = {
                browserName: 'chrome',
                logName: 'logName',
                deviceName: 'deviceName',
                platformName: 'platformName',
            } as WebdriverIO.Capabilities
            expect(await getInstanceData(capabilities, browser)).toMatchSnapshot()
        })

        it('should return platformName based on the currentCapabilities.platform', async() => {
            const capabilities = {}
            expect(await getInstanceData(capabilities, {
                ...browser,
                platform: 'browser.platform'
            } as any)).toMatchSnapshot()
        })

        it('should return correct instance data when JSONWP Mobile capabilities are provided', async() => {
            const capabilities = {
                browserName: 'chrome',
                logName: 'logName',
                deviceName: 'deviceName',
                platformName: 'platformName',
                nativeWebScreenshot: true,
            } as WebdriverIO.Capabilities
            expect(await getInstanceData(capabilities, browser)).toMatchSnapshot()
        })

        it('should return correct instance data when W3C Mobile capabilities are provided', async () => {
            const capabilities = {
                browserName: 'chrome',
                logName: 'logName',
                'appium:deviceName': 'appium:deviceName',
                platformName: 'platformName',
                'appium:nativeWebScreenshot': true,
            } as WebdriverIO.Capabilities
            expect(await getInstanceData(capabilities, browser)).toMatchSnapshot()
        })
    })
})
