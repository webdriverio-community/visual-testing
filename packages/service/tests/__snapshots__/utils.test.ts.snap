// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`utils > getFolders > should be able to return the correct folders when method options are provided 1`] = `
{
  "actualFolder": "methodActual",
  "baselineFolder": "methodBase",
  "diffFolder": "methodDiff",
}
`;

exports[`utils > getFolders > should be able to return the correct folders when no method options are provided 1`] = `
{
  "actualFolder": "folderActual",
  "baselineFolder": "folderBase",
  "diffFolder": "folderDiff",
}
`;

exports[`utils > getInstanceData > should return correct instance data when JSONWP Mobile capabilities are provided 1`] = `
{
  "appName": "not-known",
  "browserName": "chrome",
  "browserVersion": "not-known",
  "deviceName": "",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 0,
    "width": 0,
  },
  "isAndroid": undefined,
  "isIOS": undefined,
  "isMobile": undefined,
  "logName": "",
  "name": "",
  "nativeWebScreenshot": false,
  "platformName": "platformname",
  "platformVersion": "not-known",
}
`;

exports[`utils > getInstanceData > should return correct instance data when W3C Mobile capabilities are provided 1`] = `
{
  "appName": "not-known",
  "browserName": "chrome",
  "browserVersion": "not-known",
  "deviceName": "appium:devicename",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 0,
    "width": 0,
  },
  "isAndroid": undefined,
  "isIOS": undefined,
  "isMobile": undefined,
  "logName": "",
  "name": "",
  "nativeWebScreenshot": true,
  "platformName": "platformname",
  "platformVersion": "not-known",
}
`;

exports[`utils > getInstanceData > should return instance data when all capabilities are provided 1`] = `
{
  "appName": "not-known",
  "browserName": "chrome",
  "browserVersion": "not-known",
  "deviceName": "",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 0,
    "width": 0,
  },
  "isAndroid": undefined,
  "isIOS": undefined,
  "isMobile": undefined,
  "logName": "",
  "name": "",
  "nativeWebScreenshot": false,
  "platformName": "platformname",
  "platformVersion": "not-known",
}
`;

exports[`utils > getInstanceData > should return instance data when log name is provided 1`] = `
{
  "appName": "not-known",
  "browserName": "chrome",
  "browserVersion": "not-known",
  "deviceName": "",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 0,
    "width": 0,
  },
  "isAndroid": undefined,
  "isIOS": undefined,
  "isMobile": undefined,
  "logName": "",
  "name": "",
  "nativeWebScreenshot": false,
  "platformName": "osx",
  "platformVersion": "not-known",
}
`;

exports[`utils > getInstanceData > should return instance data when the appium log name is provided 1`] = `
{
  "appName": "not-known",
  "browserName": "chrome",
  "browserVersion": "not-known",
  "deviceName": "",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 0,
    "width": 0,
  },
  "isAndroid": undefined,
  "isIOS": undefined,
  "isMobile": undefined,
  "logName": "",
  "name": "",
  "nativeWebScreenshot": false,
  "platformName": "osx",
  "platformVersion": "not-known",
}
`;

exports[`utils > getInstanceData > should return instance data when the minimum of capabilities is provided 1`] = `
{
  "appName": "not-known",
  "browserName": "chrome",
  "browserVersion": "not-known",
  "deviceName": "",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 0,
    "width": 0,
  },
  "isAndroid": undefined,
  "isIOS": undefined,
  "isMobile": undefined,
  "logName": "",
  "name": "",
  "nativeWebScreenshot": false,
  "platformName": "osx",
  "platformVersion": "not-known",
}
`;

exports[`utils > getInstanceData > should return instance data when the sauce log name is provided 1`] = `
{
  "appName": "not-known",
  "browserName": "chrome",
  "browserVersion": "not-known",
  "deviceName": "",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 0,
    "width": 0,
  },
  "isAndroid": undefined,
  "isIOS": undefined,
  "isMobile": undefined,
  "logName": "",
  "name": "",
  "nativeWebScreenshot": false,
  "platformName": "osx",
  "platformVersion": "not-known",
}
`;

exports[`utils > getInstanceData > should return instance data when wdio-ics option log name is provided 1`] = `
{
  "appName": "not-known",
  "browserName": "chrome",
  "browserVersion": "not-known",
  "deviceName": "",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 0,
    "width": 0,
  },
  "isAndroid": undefined,
  "isIOS": undefined,
  "isMobile": undefined,
  "logName": "wdio-ics-logName",
  "name": "",
  "nativeWebScreenshot": false,
  "platformName": "osx",
  "platformVersion": "not-known",
}
`;

exports[`utils > getInstanceData > should return platformName based on the currentCapabilities.platform 1`] = `
{
  "appName": "not-known",
  "browserName": "chrome",
  "browserVersion": "not-known",
  "deviceName": "",
  "devicePixelRatio": 1,
  "deviceScreenSize": {
    "height": 0,
    "width": 0,
  },
  "isAndroid": undefined,
  "isIOS": undefined,
  "isMobile": undefined,
  "logName": "",
  "name": "",
  "nativeWebScreenshot": false,
  "platformName": "osx",
  "platformVersion": "not-known",
}
`;
